test/e2e/suite/flaky_code.go

```diff
@@ -0,0 +1,150 @@
+package suite
+
+import (
+	"context"
+	"fmt"
+	"slices"
+	"sync"
+	"time"
+
+	"github.com/go-logr/logr"
+	. "github.com/onsi/ginkgo/v2"
+	"github.com/onsi/ginkgo/v2/types"
+	"github.com/onsi/gomega"
+	. "github.com/onsi/gomega"
+)
+
+// FlakyCode can be used to run a test block multiple times until it succeeds.
+// It will retry the test block the number of times specified by flakeAttempts.
+// If the test block fails, the deferred cleanups will be run.
+// If the test block succeeds, the deferred cleanups will not be run until the end of the entire test.
+type FlakyCode struct {
+	Logger logr.Logger
+}
+type FlakeRun struct {
+	DeferCleanup    func(func(context.Context), ...interface{})
+	RetryableExpect func(actual interface{}, extra ...interface{}) Assertion
+}
+
+type retryable struct {
+	panicableError error
+	retryableError error
+	shouldRetry    bool
+	testGoMega     gomega.Gomega
+}
+
+func newRetryable(attempt int, maxAttempts int) *retryable {
+	retry := &retryable{}
+	retry.testGoMega = gomega.NewGomega(func(message string, callerSkip ...int) {
+		retry.shouldRetry = attempt < maxAttempts-1
```

Comment:
not a big deal but i think this makes it more clear that `shouldRetry` doesn't ever change in this struct
```suggestion
	retry := &retryable{
	    shouldRetry = attempt < maxAttempts-1
	}
	retry.testGoMega = gomega.NewGomega(func(message string, callerSkip ...int) {
```